src/OpenFeatureSDK/OpenFeature.cs

```diff
@@ -30,19 +36,53 @@ public sealed class OpenFeature
         /// Sets the feature provider
         /// </summary>
         /// <param name="featureProvider">Implementation of <see cref="FeatureProvider"/></param>
-        public void SetProvider(FeatureProvider featureProvider) => this._featureProvider = featureProvider;
+        public void SetProvider(FeatureProvider featureProvider)
+        {
+            this._featureProviderLock.EnterWriteLock();
+            try
+            {
+                this._featureProvider = featureProvider;
+            }
+            finally
+            {
+                this._featureProviderLock.ExitWriteLock();
+            }
+        }
 
         /// <summary>
         /// Gets the feature provider
+        /// <para>
+        /// The feature provider may be set from multiple threads, when accessing the global feature provider
+        /// it should be accessed once for an operation, and then that reference should be used for all dependent
+        /// operations. For instance, during an evaluation the flag resolution method, and the provider hooks
+        /// should be accessed from the same reference, not two independent calls to
+        /// <see cref="OpenFeature.GetProvider"/>.
+        /// </para>

```

Comment:
One thing I am not calling out here is that setting hooks, provider, and context are not guaranteed to be ordered among themselves. If you have one thread setting all 3, and another thread doing evaluations, then a single evaluation may use some combination of the values being set.

Reply:
> One thing I am not calling out here is that setting hooks, provider, and context are not guaranteed to be ordered among themselves. If you have one thread setting all 3, and another thread doing evaluations, then a single evaluation may use some combination of the values being set.

That makes sense, and I think mentioning this might be adding cognitive overload with not-too-much value returned.

As for the block comment, I had added something similar in the read-lock block in Java, so it makes sense to me and I think it's worth pointing out.

Reply:
So, here each of the 3 things are individually thread safe. 

The provider will use a stable reference, which covers the provider hooks and other data.

The context is immutable, and we get a reference at a specific point in time. The reference globally may change, but we are decoupled from that because we got a reference. Locking for the duration of getting the context and the provider wouldn't prevent interleaved setting of them. You could set one, and then the evaluation gets the lock, you have to wait, and then set the other. 

The last bit are the other two sets of hooks. These are interesting. There isn't anything which inherently couples them to anything else. They are copied, so the list will be stable in the places they are iterate, but up until they are iterated, they could change.

If you have cross dependent hooks then they should be added with AddHooks, which will add them atomically. I should document that.

Reply:
I added a couple comments to the hook methods.

Reply:
> If you have cross dependent hooks

Worth saying I generally think this should be avoided, but I think people might do it.

Reply:
> Locking for the duration of getting the context and the provider wouldn't prevent interleaved setting of them.

Wouldn't read-locking in the evaluation on a shared lock which is write-locked by both the provider and context mutators address exactly this? Or am I misunderstanding?

Reply:
Not unless you also locked for the duration of setting both of them.

If they can be independently set, as they can in the implementation at the moment, then it doesn't.

```
Thread 1:
SetProvider(provider)
  writeLock(_provider)
// Evaluation could start here
SetContext(context)
  writeLock(_context)

Thread 2:
Evaluate()
  readLock(_provider)
    readLock(_context)
```

So, given this code someone could intend to set the provider and context, but it would be possible for evaluation to grab the lock before thread 1 set the context, but after it set the provider.

In order to not interleave it would need to be more like

```
Thread 1:
writeLock(_provider)
  writeLock(_context)
    setProvider
    setContext
    
Thread 2:
Evaluate()
  readLock(_provider)
    readLock(_context)
```

Or use a single lock

Reply:
Ahh, yes, I agree. I misunderstood your initial statement.